backend/auth/role_map_repository.go

```diff
@@ -0,0 +1,222 @@
+package auth
+
+import (
+	"errors"
+	"log"
+	"os"
+	"sync"
+
+	"github.com/ZPI-2024-25/KubernetesAccessManager/cluster"
+	"github.com/ZPI-2024-25/KubernetesAccessManager/models"
+	"gopkg.in/yaml.v2"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type RoleMapRepository struct {
+	// Subrole map is optional, if not provided, subroles will be ignored. Roles received in token are checked only with RoleMap
+	RoleMap map[string]*models.Role
+	SubroleMap map[string]*models.Role
+}
+
+var (
+	instance      *RoleMapRepository
+	once          sync.Once
+)
+
+func GetInstance() (*RoleMapRepository, error) {
+	once.Do(func() {
+		roleMapNamespace := os.Getenv("ROLEMAP_NAMESPACE")
+		if roleMapNamespace == "" {
+			log.Printf("ROLEMAP_NAMESPACE not set, using default namespace")
+			roleMapNamespace = "default"
+		}
+
+		roleMapName := os.Getenv("ROLEMAP_NAME")
+		if roleMapName == "" {
+			log.Printf("ROLEMAP_NAME not set, using default name")
+			roleMapName = "role-mapper"
+		}
+
+		roleMap, subroleMap := GetRoleMapConfig(roleMapNamespace, roleMapName)
+		if roleMap == nil {
+			log.Printf("Failed to initialize RoleMapRepository")
+			return
+		}
+		instance = &RoleMapRepository{RoleMap: roleMap, SubroleMap: subroleMap}
+
+		log.Printf("RoleMapRepository initialized with %d roles %d subroles", len(instance.RoleMap), len(instance.SubroleMap))
+	})
+
+	if instance == nil {
+		return nil, errors.New("failed to initialize RoleMapRepository")
+	}
+
+	return instance, nil
+}
+
+func (rmr *RoleMapRepository) HasPermission(rolenames []string, operation *models.Operation) bool {
+	visited := make(map[string]interface{})
+	for _, role := range rolenames {
+		if rmr.hasPermission(role, operation, visited) {
+			return true
+		}
+	}
+	return false
+}
```

Comment:
It could be a function, but you would have to pass the RoleMapRepository to it anyways for testing purposes. I think this it is better as method however as it is inherently dependant on the repositoryMap state and actually a matter of preference.